dg.go

```diff
@@ -1,315 +1,480 @@
+	for fromNodeID := range n.dg.connectionsToNode[n.id] {
+		delete(n.dg.connectionsFromNode[fromNodeID], n.id)
+	}
+	delete(n.dg.connectionsToNode, n.id)
+	delete(n.dg.nodes, n.id)
+	n.deleted = true
+	return nil
 }
 
-func (d *node[NodeType]) DisconnectInbound(fromNodeID string) error {
-    d.dg.lock.Lock()
-    defer d.dg.lock.Unlock()
-    if d.deleted {
-        return &ErrNodeDeleted{
-            d.id,
-        }
-    }
-    if _, ok := d.dg.nodes[fromNodeID]; !ok {
-        return &ErrNodeNotFound{
-            fromNodeID,
-        }
-    }
-    if _, ok := d.dg.connectionsToNode[d.id][fromNodeID]; !ok {
-        return &ErrConnectionDoesNotExist{
-            d.id,
-            fromNodeID,
-        }
-    }
-    delete(d.dg.connectionsToNode[d.id], fromNodeID)
-    delete(d.dg.connectionsFromNode[fromNodeID], d.id)
-    return nil
+func (n *node[NodeType]) ListInboundConnections() (map[string]Node[NodeType], error) {
+	n.dg.lock.Lock()
+	defer n.dg.lock.Unlock()
+	if n.deleted {
+		return nil, &ErrNodeDeleted{n.id}
+	}
+	result := make(map[string]Node[NodeType], len(n.dg.connectionsToNode[n.id]))
+	for fromNodeID := range n.dg.connectionsToNode[n.id] {
+		result[fromNodeID] = n.dg.nodes[fromNodeID]
+	}
+	return result, nil
 }
 
-func (d *node[NodeType]) DisconnectOutbound(toNodeID string) error {
-    d.dg.lock.Lock()
-    defer d.dg.lock.Unlock()
-    if d.deleted {
-        return &ErrNodeDeleted{
-            d.id,
-        }
-    }
-    if _, ok := d.dg.nodes[toNodeID]; !ok {
-        return &ErrNodeNotFound{
-            toNodeID,
-        }
-    }
-    if _, ok := d.dg.connectionsFromNode[d.id][toNodeID]; !ok {
-        return &ErrConnectionDoesNotExist{
-            d.id,
-            toNodeID,
-        }
-    }
-    delete(d.dg.connectionsFromNode[d.id], toNodeID)
-    delete(d.dg.connectionsToNode[toNodeID], d.id)
-    return nil
+func (n *node[NodeType]) ListOutboundConnections() (map[string]Node[NodeType], error) {
+	n.dg.lock.Lock()
+	defer n.dg.lock.Unlock()
+	if n.deleted {
+		return nil, &ErrNodeDeleted{n.id}
+	}
+	result := make(map[string]Node[NodeType], len(n.dg.connectionsFromNode[n.id]))
+	for toNodeID := range n.dg.connectionsFromNode[n.id] {
+		result[toNodeID] = n.dg.nodes[toNodeID]
+	}
+	return result, nil
 }
 
-func (d *node[NodeType]) Remove() error {
-    d.dg.lock.Lock()
-    defer d.dg.lock.Unlock()
-    if d.deleted {
-        return &ErrNodeDeleted{
-            d.id,
-        }
-    }
-    for toNodeID := range d.dg.connectionsFromNode[d.id] {
-        delete(d.dg.connectionsToNode[toNodeID], d.id)
-    }
-    delete(d.dg.connectionsFromNode, d.id)
-    for fromNodeID := range d.dg.connectionsToNode[d.id] {
-        delete(d.dg.connectionsFromNode[fromNodeID], d.id)
-    }
-    delete(d.dg.connectionsToNode, d.id)
-    delete(d.dg.nodes, d.id)
-    d.deleted = true
-    return nil
+// dependencyResolved is used to notify a node that one of its dependencies have had their resolution
```

Comment:
"One" of "its" dependencies have had "their" status set? Can we match the prepositions here?

Reply:
Changed.